Improve Synchronization UI

runkaiz opened the pull request:
Partially implements #837.

This is my first pull-request (well, first time attempting to contribute to a public open source project in fact). Code changes are minimum so please give me some advice if I am doing something wrong here, any help would be appreciated, just trying to gain some experience.

Neui commented:
Not entirely sure if you want some feedback now, but:
* You hide the button, but never make it `.show()` it again (at least from what I can see from the changes, I haven't tested the change). There is [`.set_visible()`](https://lazka.github.io/pgi-docs/#Gtk-3.0/classes/Widget.html#Gtk.Widget.set_visible) you can use with an boolean parameter. if you want to make it a one-liner.
* `print()` is OK when debugging, just make sure to remove them before committing/submitting. (I'd say in a draft PR it would be OK, since code is likely to be changed anyway). You could also use the logging system to print stuff out like `log.debug("xyz")` if you think it could be useful later when debugging (you need to `import logging` and `log = logging.getLogger(__name__)` near the top of the file, like for example https://github.com/getting-things-gnome/gtg/blob/5f72c8a9319e06d42967fb8461ea832e8cd095d9/GTG/gtk/browser/adaptive_button.py#L1-L8

For the link stuff (mentioned in the linked issue), https://docs.gtk.org/gtk3/class.Label.html#links would be a start.

runkaiz commented:
I believe most of not all of the requested changes have been applied. I might have misinterpreted the part about hiding the enable / disable button on the default back-end so I would appreciate it if someone can help me review this.

Neui commented on GTG/gtk/backends/addpanel.py:
```diff
@@ -170,7 +172,16 @@ def on_combo_changed(self, widget=None):
         backend_name = self.combo_types.get_selected()
         if backend_name is None:
+            if 'backend_caldav' in inactive_modules:
+                markup = '<big>Error: Python package \'caldev\' not installed.</big>'
```
```suggestion
                markup = '<big>{}</big>'.format(_('Error: Python package \'caldav\' not installed.'))
```
Typo in `caldav`, and make it translatable.

Neui commented on GTG/gtk/backends/addpanel.py:
```diff
@@ -170,7 +172,16 @@ def on_combo_changed(self, widget=None):
+                return
+
+
+            markup = '<big>Error: An unknown backend could not be loaded.</big>'
```
```suggestion
            markup = '<big>{}</big>'.format(_('Error: An unknown backend could not be loaded.'))
```

Make it translatable.

Neui commented on GTG/backends/__init__.py:
```diff
@@ -67,10 +67,12 @@ def __init__(self):
                 # Something is wrong with this backend, skipping
                 log.warning("Backend %s could not be loaded: %r",
                             module_name, exception)
+                inactive_modules.append(module_name)
```
```python-traceback
Traceback (most recent call last):
  File "/app/lib/python3.9/site-packages/GTG/backends/__init__.py", line 65, in __init__
    __import__(extended_module_name)
  File "/app/lib/python3.9/site-packages/GTG/backends/backend_caldav.py", line 30, in <module>
    import caldav
ModuleNotFoundError: No module named 'caldav'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/lib/python3.9/site-packages/GTG/gtk/application.py", line 124, in do_startup
    for backend_dic in BackendFactory().get_saved_backends_list():
  File "/app/lib/python3.9/site-packages/GTG/backends/__init__.py", line 70, in __init__
    inactive_modules.append(module_name)
AttributeError: 'set' object has no attribute 'append'
```

You need to use [`.add()`](https://docs.python.org/3.8/library/stdtypes.html#frozenset.add) since it is a set, not an list.

Neui commented on GTG/backends/__init__.py:
```diff
@@ -67,10 +67,12 @@ def __init__(self):
             except Exception:
                 # Other exception log as errors
                 log.exception("Malformated backend %s:", module_name)
+                inactive_modules.append(module_name)
```
Ditto. You need to use [`.add()`](https://docs.python.org/3.8/library/stdtypes.html#frozenset.add) since it is a set, not an list.

Neui commented on GTG/backends/__init__.py:
```diff
@@ -32,7 +32,7 @@
 
 log = logging.getLogger(__name__)
-
+inactive_modules = set()
```
Might be better to move this to the `BackendFactory` class below. The [`Borg`](https://github.com/getting-things-gnome/gtg/blob/5f72c8a9319e06d42967fb8461ea832e8cd095d9/GTG/core/borg.py) class it inherits is basically making it a singleton (state shared on every instance).

Neui reviewed (changes requested):
I am not sure if we should make a link to the XML file. At least for me, clicking on it does nothing, and if something did, it may encourage people to modify the file while GTG is still open.

Images would have been nice to look at. Just launch `gnome-screenshot -i`, select region, copy to clipboard and paste inside GitHub textbox thing.

![Default](https://user-images.githubusercontent.com/1196130/170894755-f33a9cab-870d-4df5-8c74-76c22b01a78d.png)

After fixing the `.append()` issue:
![caldav not found](https://user-images.githubusercontent.com/1196130/170895148-b7d5f1cf-911c-4741-bd3d-d23528021acb.png)

I don't quite like how it looks like. Maybe allow Caldav to be selectable, but name it like "caldav (error)" or something, so if you have multiple backends you can see which ones have an error.

nekohayo commented:
The big refactoring has finally landed (see the pinned 0.7 tracking issue for details)!

Please rebase this to the latest `master`. Apologies for the inconvenience; the new codebase is expected to be much better to work with.

nekohayo commented:
Once this gets rebased, I would like @jaesivsm to help reviewing this.

pentux-GitHub commented:
The need is still real, #837 is still open. The blocker is that this predates the GTK4 port and touches backend UI files reworked since, so it needs a rebase, as @nekohayo asked back in 2024.

@runkaiz if you'd like to rebase it on current master, it'd be welcome; otherwise it could be closed and #837 left open for a fresh attempt.

runkaiz commented:
Alright I am on it.

runkaiz committed:
backends dialog: explain why no service can be added

When no synchronization service is available (for example when
python-caldav is not installed), the add panel used to show an empty
combo box, blank labels and an OK button that did nothing.

Instead, hide the empty combo box, disable the OK button and show a
human-readable explanation listing which backend modules could not be
loaded and why, based on BackendFactory.get_inactive_modules().

Since an empty combo box never emits 'changed', refresh_backends() now
also calls on_combo_changed() explicitly so the explanation is shown.

Part of #837

runkaiz committed:
backends: track modules that fail to load in BackendFactory

Record the name and error message of every backend module that fails to
import (or raises during import) in BackendFactory.inactive_modules,
exposed through get_inactive_modules(). This lets the UI explain why a
backend is unavailable, e.g. when the python 'caldav' package is not
installed.

Part of #837

runkaiz committed:
backends dialog: hide the sync toggle for the default backend

The default backend cannot be disabled, so hide the Enable/Disable
syncing button instead of showing it permanently greyed out.

Part of #837

runkaiz commented:
Rebased onto current `master` (the three old commits are replaced by a fresh implementation on top of the new codebase).

Since the GTK4 refactor removed `backend_localfile`, parts of the original approach no longer applied, so the scope is now:

- **`BackendFactory` records failed modules**: `inactive_modules` maps each backend module that failed to import to its error message, exposed via `get_inactive_modules()`. This also fixes the `set`/`.append()` mix-up @Neui caught in the original review.
- **Add panel explains the empty state**: with no loadable service (e.g. python-caldav not installed), the panel used to show an empty combo box, blank labels, and a no-op OK button. It now hides the combo, disables OK, and lists which modules failed and why — generically, not hardcoded to caldav. One subtlety: an empty combo never emits `changed`, so `refresh_backends()` now calls `on_combo_changed()` explicitly, otherwise the explanation would never render.
- **Configure panel hides (rather than greys out) the Enable/Disable syncing button** for the default backend, per #837.

Dropped from the original PR based on review feedback: the hyperlink to the XML data file (it also relied on a hardcoded Flatpak path) and the caldav-specific error string.

Verified locally on a machine without python-caldav (the exact #837 scenario): the factory reports `{'backend_caldav': "No module named 'caldav'"}` and the add panel renders the explanation with the OK button disabled.

Written by AI.

runkaiz commented:
Definitely need a live run from someone, I do not have a live linux environment at the moment. Compilation and tests passes.

pentux-GitHub commented:
Live run done on Linux with the bryce sample dataset, on top of current master (merges cleanly, no conflict with the recent CalDAV commits).

Nominal case: the add dialog lists CalDAV as before, no visible regression. Simulated missing dependency (caldav import blocked): the dialog now shows "No synchronization services available" with the failing module and the reason, OK disabled. That is exactly what #837 asked for and it works as described.

One honest side note: a few GTK warnings show up in the console when the empty state is displayed (a GtkStack child name lookup and two negative size reports). I have not established whether they come from this branch or were already there, and they don't affect the behavior.

From my side this looks ready for review. The CI still needs a maintainer to approve the workflow run on this fork.

pentux-GitHub commented:
The rebase requested above is done and live tested (previous comments). Two things remain: the CI workflow on this fork still needs a maintainer approval to run, and nekohayo wrote earlier that once rebased he would like @jaesivsm to review. The rebase condition is now met, so pinging you here, when you have a moment.
